sys/include/net/netif.h

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright (C) 2016 Kaspar Schleiser <kaspar@schleiser.de>
+ * Copyright (C) 2017 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    net_netif   Network interfaces
+ * @ingroup     net
+ * @brief       Common network interface API
+ * @{
+ *
+ * @file
+ * @brief
+ *
+ * @author  Martine Lenders <m.lenders@fu-berlin.de>
+ * @author  Kaspar Schleiser <kaspar@schleiser.de>
+ */
+#ifndef NETIF_H_
+#define NETIF_H_
+
+#include "net/netdev.h"
+#include "net/netopt.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Identifier for an invalid network interface
+ */
+#define NETIF_INVALID       (0x0-1)
+
+/**
+ * @brief   Maximum length for an interface name
+ */
+#ifndef NETIF_NAMELENMAX
+#define NETIF_NAMELENMAX    (8U)
+#endif
+
+/**
+ * @brief   Type of the interface
+ */
+typedef enum {
+    NETIF_TYPE_INVALID = 0,     /**< Interface is of no specific type (should not be used) */
+#if defined(MODULE_GNRC) || DOXYGEN
+    /**
+     * @brief   Default GNRC network interface (without software MAC)
+     *
+     * @note only available with @ref net_gnrc "GNRC" module.
+     */
+    NETIF_TYPE_GNRC_DEFAULT,
+#endif
+#if defined(MODULE_LWIP_ETHERNET) || DOXYGEN
+    /**
+     * @brief   lwIP network interface
+     *
+     * @note only available with @ref pkg_lwip.
+     */
+    NETIF_TYPE_LWIP,
+#endif /* MODULE_LWIP */
+#if defined(MODULE_EMB6) || DOXYGEN
+    /**
+     * @brief   emb6 interface
+     *
+     * @note only available with @ref pkg_emb6
+     */
+    NETIF_TYPE_EMB6,
+#endif
+    /* extend list if needed */
+} netif_type_t;
+
+/**
+ * @brief struct holding all parameters needed for interface initialization
+ */
+typedef struct {
+    /**
+     * @brief Network device parameters
+     *
+     * The network device parameters are used to associate a network device to
+     * the network interface.
+     */
+    const void *dev_params;
+    netif_type_t type;      /**< type of the interface */
+    const char *name;       /**< name of the interface */
```

Comment:
Hi, I think that makes the initialization just more complicated, because then we need to keep track of how many interfaces are already initialized. The auto-initialization is already pretty complicated so let's not make things more complicated for a totally *optional* thing for most stacks (the only stack actually requiring a name at the moment is lwIP).